**Data preparation:**
*   Gather all dog images from here: https://www.kaggle.com/c/dog-breed-identification

*   Gather all cat images from here: https://www.kaggle.com/c/dogs-vs-cats

*  Divide by dog breeds into two sets: training and test.
Note: breeds in the test set cannot appear in the training set.

* Divide the cat images also in train and test

In [4]:
import os

In [ ]:
import os
from google.colab import drive
#!pip install kaggle

In [ ]:
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!mkdir '/content/drive/MyDrive/Dataset/Dogs vs Cats'


In [ ]:
! mkdir ~/.kaggle

In [ ]:
# Set the path to your Kaggle API token on Google Drive
kaggle_token_path = '/content/drive/MyDrive/Colab Notebooks/kaggle.json'

# Set the path to save Kaggle data in Google Drive
kaggle_data_path = '/content/drive/MyDrive/Dataset/Dogs vs Cats'

# Copy the Kaggle API token to the appropriate location in Colab
!cp '{kaggle_token_path}' ~/.kaggle/kaggle.json

In [ ]:
# Set appropriate permissions for the API token
!chmod 600 ~/.kaggle/kaggle.json

In [ ]:
# Download Dogs vs Cats dataset using the Kaggle API
!kaggle competitions download -c dogs-vs-cats -p '{kaggle_data_path}'

# Download Dog Breed Identification dataset using the Kaggle API
!kaggle competitions download -c dog-breed-identification -p '{kaggle_data_path}'

 99% 806M/812M [00:08<00:00, 142MB/s]
100% 812M/812M [00:08<00:00, 104MB/s]
100% 691M/691M [00:09<00:00, 82.0MB/s]
100% 691M/691M [00:09<00:00, 75.9MB/s]


In [ ]:
import zipfile

# Extract Dogs vs Cats dataset
with zipfile.ZipFile('/content/drive/MyDrive/Dataset/Dogs vs Cats/dogs-vs-cats.zip', 'r') as zip_ref:
    zip_ref.extractall('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats')

# Extract Dog Breed Identification dataset
with zipfile.ZipFile('/content/drive/MyDrive/Dataset/Dogs vs Cats/dog-breed-identification.zip', 'r') as zip_ref:
    zip_ref.extractall('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog Breed Identification')

#removing the zip files
os.remove('/content/drive/MyDrive/Dataset/Dogs vs Cats/dog-breed-identification.zip')
os.remove('/content/drive/MyDrive/Dataset/Dogs vs Cats/dogs-vs-cats.zip')


In [ ]:
# Not Extracting test1.zip as these images are not labeled
# Extracting train.zip only

with zipfile.ZipFile('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats/train.zip', 'r') as zip_ref:
    zip_ref.extractall('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats')


In [ ]:
#removing the zip files and sample submission files
os.remove('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats/test1.zip')
os.remove('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats/train.zip')
os.remove('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats/sampleSubmission.csv')
os.remove('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog Breed Identification/sample_submission.csv')


In [ ]:
!mkdir '/content/drive/MyDrive/Dataset/Dogs vs Cats/Cats'

In [ ]:
import os
import shutil

# Path to the new folder for storing cat images
cat_folder = '/content/drive/MyDrive/Dataset/Dogs vs Cats/Cats'

# Loop through the training cat images and move them to the new folder
for filename in os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats/train'):
    if filename.startswith('cat'):
        src = os.path.join('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats/train', filename)
        dst = os.path.join(cat_folder, filename)
        shutil.copy(src, dst)


In [ ]:
len(os.listdir(cat_folder))


12500

In [ ]:
len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog Breed Identification/train'))

10222

In [ ]:
!mkdir '/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_train' #for keeping 105 breeds

In [ ]:
!mkdir '/content/drive/MyDrive/Dataset/Dogs vs Cats/DogTest'    #for keeping 15 breeds

In [ ]:
import pandas as pd
labels_df=pd.read_csv('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog Breed Identification/labels.csv')
labels_df

id                     breed
0      000bec180eb18c7604dcecc8fe0dba07               boston_bull
1      001513dfcb2ffafc82cccf4d8bbaba97                     dingo
2      001cdf01b096e06d78e9e5112d419397                  pekinese
3      00214f311d5d2247d5dfe4fe24b2303d                  bluetick
4      0021f9ceb3235effd7fcde7f7538ed62          golden_retriever
...                                 ...                       ...
10217  ffd25009d635cfd16e793503ac5edef0                    borzoi
10218  ffd3f636f7f379c51ba3648a9ff8254f            dandie_dinmont
10219  ffe2ca6c940cddfee68fa3cc6c63213f                  airedale
10220  ffe5f6d8e2bff356e9482a80a6e29aac        miniature_pinscher
10221  fff43b07992508bc822f33d8ffd902ae  chesapeake_bay_retriever

[10222 rows x 2 columns]

In [ ]:
labels_df.breed.value_counts()

scottish_deerhound      126
maltese_dog             117
afghan_hound            116
entlebucher             115
bernese_mountain_dog    114
                       ... 
golden_retriever         67
brabancon_griffon        67
komondor                 67
eskimo_dog               66
briard                   66
Name: breed, Length: 120, dtype: int64

In [ ]:
# Paths
dog_train_folder = '/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_train'
dog_test_folder = '/content/drive/MyDrive/Dataset/Dogs vs Cats/DogTest'

# Get the list of unique breeds and select 105 for training and 15 for testing
all_breeds = labels_df['breed'].unique()
breeds_for_train = all_breeds[:105]
breeds_for_test = all_breeds[105:120]

# Iterating through the labels and copy images to appropriate folders
for index, row in labels_df.iterrows():
    image_filename = f"{row['id']}.jpg"
    source_path = os.path.join('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog Breed Identification/train', image_filename)

    if row['breed'] in breeds_for_train:
        destination_path = os.path.join(dog_train_folder, image_filename)
    elif row['breed'] in breeds_for_test:
        destination_path = os.path.join(dog_test_folder, image_filename)
    else:
        continue  # Skip breeds not in selected sets

    shutil.copy(source_path, destination_path)

# Creating separate DataFrames for train and test labels (for ground truth)
train_labels_df = labels_df[labels_df['breed'].isin(breeds_for_train)]
test_labels_df = labels_df[labels_df['breed'].isin(breeds_for_test)]

train_labels_df.to_csv('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_train_labels.csv', index=False)
test_labels_df.to_csv('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_test_labels.csv', index=False)


In [ ]:
print(len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_train')))
print(len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/DogTest')))

8995
1227


In [ ]:
!mkdir '/content/drive/MyDrive/Dataset/Dogs vs Cats/Train_Dog_cat'

In [ ]:
# Copy all files from Cats folder to Train_Dog_cat folder
for filename in os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Cats'):
    src = os.path.join('/content/drive/MyDrive/Dataset/Dogs vs Cats/Cats', filename)
    dst = os.path.join('/content/drive/MyDrive/Dataset/Dogs vs Cats/Train_Dog_cat', filename)
    shutil.copy(src, dst)

# Copy all files from Dog_train folder to Train_Dog_cat folder
for filename in os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_train'):
    src = os.path.join('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_train', filename)
    dst = os.path.join('/content/drive/MyDrive/Dataset/Dogs vs Cats/Train_Dog_cat', filename)
    shutil.copy(src, dst)

In [ ]:
len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Train_Dog_cat'))

21495

In [5]:
print(len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/DogTest')))

1227


In [2]:
from google.colab import drive
drive.flush_and_unmount()
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [ ]:
import shutil
# Remove the folder and its contents
shutil.rmtree('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dogs vs Cats')
shutil.rmtree('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog Breed Identification')

In [7]:
# Create the Train folder with Cat and Dog subfolders
!mkdir -p '/content/drive/MyDrive/Dataset/Dogs vs Cats/Train/Cat'
!mkdir -p '/content/drive/MyDrive/Dataset/Dogs vs Cats/Train/Dog'

# Create the Test folder with Cat and Dog subfolders
!mkdir -p '/content/drive/MyDrive/Dataset/Dogs vs Cats/Test/Cat'
!mkdir -p '/content/drive/MyDrive/Dataset/Dogs vs Cats/Test/Dog'

In [8]:
import random
import shutil

# Source directory with cat images
source_directory = '/content/drive/MyDrive/Dataset/Dogs vs Cats/Cats'

# Destination directories for training and testing
train_directory = '/content/drive/MyDrive/Dataset/Dogs vs Cats/Train/Cat'
test_directory = '/content/drive/MyDrive/Dataset/Dogs vs Cats/Test/Cat'

# Get a list of cat image filenames
cat_image_filenames = os.listdir(source_directory)
random.shuffle(cat_image_filenames)  # Shuffle the list

# Number of images for training and testing
num_train_images = 8995
num_test_images = 3505

# Move images for training
for filename in cat_image_filenames[:num_train_images]:
    source_path = os.path.join(source_directory, filename)
    destination_path = os.path.join(train_directory, filename)
    shutil.move(source_path, destination_path)

# Move images for testing
for filename in cat_image_filenames[num_train_images:num_train_images+num_test_images]:
    source_path = os.path.join(source_directory, filename)
    destination_path = os.path.join(test_directory, filename)
    shutil.move(source_path, destination_path)


In [9]:
print(len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Train/Cat')))
print(len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Test/Cat')))

8995
3505


In [10]:
# Source directories with dog images
dog_train_directory = '/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_train'
dog_test_directory = '/content/drive/MyDrive/Dataset/Dogs vs Cats/DogTest'

# Destination directories for training and testing
train_dog_directory = '/content/drive/MyDrive/Dataset/Dogs vs Cats/Train/Dog'
test_dog_directory = '/content/drive/MyDrive/Dataset/Dogs vs Cats/Test/Dog'

# Get a list of dog image filenames
dog_train_image_filenames = os.listdir(dog_train_directory)
dog_test_image_filenames = os.listdir(dog_test_directory)

# Move dog images for training
for filename in dog_train_image_filenames:
    source_path = os.path.join(dog_train_directory, filename)
    destination_path = os.path.join(train_dog_directory, filename)
    shutil.move(source_path, destination_path)

# Move dog images for testing
for filename in dog_test_image_filenames:
    source_path = os.path.join(dog_test_directory, filename)
    destination_path = os.path.join(test_dog_directory, filename)
    shutil.move(source_path, destination_path)


In [11]:
print(len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Train/Dog')))
print(len(os.listdir('/content/drive/MyDrive/Dataset/Dogs vs Cats/Test/Dog')))

8995
1227


In [12]:
# removing unnecessary folders

shutil.rmtree('/content/drive/MyDrive/Dataset/Dogs vs Cats/Cats')
shutil.rmtree('/content/drive/MyDrive/Dataset/Dogs vs Cats/DogTest')
shutil.rmtree('/content/drive/MyDrive/Dataset/Dogs vs Cats/Dog_train')

In [13]:
shutil.rmtree('/content/drive/MyDrive/Dataset/Dogs vs Cats/Train_Dog_cat')

So, According to the specification:

- Dogs vs Cats
  - Train
  - Test

- The `Train` directory contains two subdirectories:
  - `Cat` with 8995 cat images for training.
  - `Dog` with 8995 dog images for training.

- The `Test` directory contains two subdirectories:
  - `Cat` with 3505 cat images for testing.
  - `Dog` with 1227 dog images for testing.


